Shared/Samples/Add custom dynamic entity data source/AddCustomDynamicEntityDataSourceView.Model.swift

```diff
@@ -0,0 +1,140 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import ArcGIS
+import Combine
+import Foundation
+
+extension AddCustomDynamicEntityDataSourceView {
+    /// The model used to store the geo model and other expensive objects
+    /// used in this view.
+    class Model: ObservableObject {
+        /// A map with an ArcGIS oceans basemap style.
+        let map: Map = {
+            let map = Map(basemapStyle: .arcGISOceans)
+            map.initialViewpoint = Viewpoint(
+                latitude: 47.984,
+                longitude: -123.657,
+                scale: 3e6
+            )
+            return map
+        }()
+        
+        /// The dynamic entity layer that is displaying our custom data.
+        let dynamicEntityLayer: DynamicEntityLayer
+        
+        init() {
+            // The meta data for the custom dynamic entity data source.
+            let info = DynamicEntityDataSourceInfo(
+                entityIDFieldName: Vessel.Attributes.CodingKeys.mmsi.rawValue,
+                fields: .vesselFields
+            )
+            
+            info.spatialReference = .wgs84
+            
+            /// Makes the vessel feed that emits the events to the custom dynamic entity data source.
+            /// - Returns: The vessel feed.
+            func makeFeed() throws -> VesselFeed {
+                // All of the JSON as a string.
+                let contents = try String(contentsOf: .selectedVesselsDataSource, encoding: .utf8)
+                
+                // An array of the individual JSON strings.
+                let lines = contents.split(separator: "\n")
+                
+                let decoder = JSONDecoder()
+                
+                // The feed events that the feed will emit.
+                let feedEvents = lines
+                    .compactMap { jsonString -> CustomDynamicEntityFeedEvent? in
+                        guard let decodable = try? decoder.decode(
+                            Vessel.self,
+                            from: jsonString.data(using: .utf8)!
+                        ) else { return nil }
+                        
+                        // The geometry that was decoded from the JSON.
+                        let geometry = decodable.geometry
+                        
+                        // We successfully decoded the vessel JSON so we should
+                        // add that vessel as a new observation.
+                        return .newObservation(
+                            geometry: Point(x: geometry.x, y: geometry.y, spatialReference: .wgs84),
+                            attributes: decodable.attributes.makeDictionary()
+                        )
+                    }
+                
+                return VesselFeed(events: feedEvents)
```

Comment:
Initially I thought it would be better reading the JSON file [line by line](https://stackoverflow.com/a/75882427/14369688), but it seems the `feedEvents` is a static sequence rather than an async sequence, which means all the events are in memory before the data source starts. Is it how this API supposed to work?

Reply:
I too was thinking about this. It can be done with `FileHandle` or with `URL`, but I couldn't find a way to get a publisher from an async sequence. We could still go down that route using [Throttle](https://github.com/apple/swift-async-algorithms/blob/main/Sources/AsyncAlgorithms/AsyncAlgorithms.docc/Guides/Throttle.md), but it would require adding another dependency. I don't really see a problem with that.

Reply:
Hm, I see your point. I was thinking since we already have the data available we can just load it beforehand but maybe that is not the best to show users since they will probably be receiving live data.

Reply:
> I don't really see a problem with that.

Is it worth adding a dependency, just for this one sample?

Reply:
I think so. Async sequences are the new kid on the block while Combine is aging quickly. If we anticipate that users will need to do something similar (read a large file line-by-line asynchronously), it would be great to demonstrate for them the best way to do that. Using a publisher definitely isn't, since it requires reading the whole file into memory first.

Reply:
After some reading I didn't find a good native way to "map" an async sequence to another, as pointed out by this user on SO: https://stackoverflow.com/a/73426716/14369688

So I borrowed his idea to roll our own feed with a `Task` block. But I haven't investigate the lifetime of this approach. It seems to work just fine. Bottom line, we can add a feed cancel/disconnect method if we need? 🤔 

What about sth like this?

```swift
            // call site of the custom feed struct
            let customDataSource = CustomDynamicEntityDataSource(info: info) {
                VesselFeed(inputLog: URL.selectedVesselsDataSource.lines)
            }

private struct VesselFeed: CustomDynamicEntityFeed {
    let inputLog: AsyncLineSequence<URL.AsyncBytes>
    
    var events: AsyncStream<CustomDynamicEntityFeedEvent> {
        let decoder = JSONDecoder()
        
        return AsyncStream<CustomDynamicEntityFeedEvent> { continuation in
            Task {
                for try await line in inputLog {
                    try await Task.sleep(nanoseconds: 10_000_000)
                    
                    let decodable = try decoder.decode(Vessel.self, from: line.data(using: .utf8)!)
                    
                    // The geometry that was decoded from the JSON.
                    let geometry = decodable.geometry
                    
                    // We successfully decoded the vessel JSON so we should
                    // add that vessel as a new observation.
                    continuation.yield(
                        .newObservation(
                            geometry: Point(x: geometry.x, y: geometry.y, spatialReference: .wgs84),
                            attributes: decodable.attributes.makeDictionary()
                        )
                    )
                }
                continuation.finish()
            }
        }
    }
}
```

Reply:
I just used `Task.sleep` in the for await loop when reading the file lines, which delays the observations being sent out which seems like a good middle ground here.

Reply:
Ah, that is exactly what you did too! I just saw your message.